### PR TITLE
Remove support for unrecognized keywords

### LINKED ISSUE #1340: Update spec to disallow unknown keywords
Following @handrews' discussion on [disallowing non-vocab keywords](https://github.com/orgs/json-schema-org/discussions/241), we decided that this is a feature that we want in order to help guarantee forward compability.

The spec needs to be updated accordingly.

### LINKED ISSUE #1365: Improved handling of unknown properties
There are many applications where a schema is distributed for use by numerous, arbitrary validators. For example, an API describes to clients a range of values that are acceptable as input, or to describe exception cases that require special handling. In these applications, it’s important that all validators agree on the set of instances that a particular schema accepts and rejects. (It goes without saying by “validator” I mean “compliant validator.”)

However, JSON Schema has one quirk in particular that prevent this from being possible: Unknown keywords are treated if they didn’t exist. For a constraint-based system, this is a serious problem when a new keyword is introduced. If a schema author uses a newly standardized validation keyword, it’s likely for an important reason that should not be ignored—which is what older validators will do. While it should be possible to write keywords that are never ignored, this is not actually possible.

There’s two obvious ways to support this application; one is to introduce a scheme so that some properties in an object can be designated “must-understand” and require validators to implement this. The other is to report/handle keywords that are unknown, so the application can react appropriately.

The $vocabulary and $schema keywords offer partial solutions of the first class, but they are not complete. These keywords are a level of abstraction removed from the keywords the schema actually requires for correct processing; as a consequence, in the worst case, you can have two validators that cannot handle any schemas in common: you may have situations where two clients cannot be supported at the same time, even if their behavior is the same.

It seems that two schemas which have the same behavior, should also be supported by the same validators. So a general solution to this problem is likely to be a of the second class: Don’t _require_ unknown keywords be ignored; but define some option to handle them unless they’re known to not be validation keywords (i.e. they are known to never reject an instance).

That is, the specification should describe how to identify those keywords that are unknown; and describe special cases where unknown keywords should not be ignored, but instead can be reported, raise an exception, or (if the details are unimportant) simply reject. Most validators can suffice with a runtime option: "Unknown keyword behavior: accept/error/reject". (The behavior currently required of implementations would be the "accept" option.)

I carefully word it this way so that, at least for now, there is no change in how schemas are processed (by default); rather this accommodates a class of applications that can enable stricter handling when that's what the application requires, and the impact of setting “error” as the default behavior can be better analyzed. (Though this can be decided now too, if some need presents itself.)

---

This is the first in a series of changes to implement [JSON Schema Extensions](https://github.com/orgs/json-schema-org/discussions/257), including:

* Standard way to indicate annotation-only keywords
* Refine definition of keywords, arguments, validation
* Typed “format” validation keywords
* Error handling keywords (to force rejection even by validators that ignore & accept unknown keywords)

It’s possible It may be preferable to implement some of these first, since that would lower the potential impact of this change.

### PR DESCRIPTION
**Pull request by gregsdennis**
We discussed this a lot, but never actually made the changes.

Resolves #1340
Resolves #1365

### CODE REVIEW

File: jsonschema-core.md
Comment (Relequestual) [inline jsonschema-core.md:980]: This section is now contradictory.

> SHOULD proceed with processing the schema.

But considered unrecognized keywords...

> mplementations MUST refuse to process schemas which contain keywords they do not recognize, or that they recognize but do not support.

Are we really wanting to say that optinal support of a vocabulary is not something we want to support anymore?

**Review by Relequestual — CHANGES_REQUESTED** [review]
Left a comment. Currently contradictory as far as I can tell.

Otherwise, looks good.

File: jsonschema-core.md
Comment (gregsdennis) [reply jsonschema-core.md:980]: Yes. This is a contradiction, which I've brought up in a discussion elsewhere (will provide the link shortly).

File: jsonschema-core.md
Comment (jdesrosiers) [inline jsonschema-core.md:438]: The wording here _feels_ contradictory although I get that it isn't and why it isn't. I get that `x-` keywords are considered "recognized", but I'm not sure we're making that clear. Maybe we need to be more clear what is considered "recognized" or maybe we need to say it different way that's harder to misunderstand.

Comment (gregsdennis) [reply jsonschema-core.md:438]: @Relequestual had this same concern in [Slack](https://json-schema.slack.com/archives/CT7FF623C/p1717978845635409).  It's addressed in #1518.

File: jsonschema-core.md
Comment (gregsdennis) [reply jsonschema-core.md:70]: > The `x-` convention is an extension mechanism...

I'm curious of your definition of an extension mechanism.  The `x-` behavior is defined by the spec, so in my eyes, these keywords are not extensions.  This is clarified by #1518.

(This paragraph is going to change in #1510 anyway, so I'm not too concerned about getting it 100% right here.)

Comment (jdesrosiers) [reply jsonschema-core.md:70]: I think we agree that the vocabulary system is an extension mechanism. It allows you to use custom keywords in your schemas. But, we decided that the vocabulary system was too cumbersome for users who just wanted to use simple annotation custom keywords, so we introduced `x-` as an alternative for users to define simple custom keywords without the overhead of the vocabulary system. Both are extension mechanisms for using custom keywords. One is complicated and powerful and the other is simple and highly constrained, but they're serving the same purpose.

`x-` is so constrained that I can see how you can view it as one thing defined by the spec, but I think there's more to it. `x-foo` and `x-bar` have the same annotation collection and validation behavior, but they aren't the same keyword. They have different semantics given to them by the schema author, not by the spec. I see the spec defined behavior of `x-` as constraints that `x-` extension keywords must adhere to, but it's not a complete definition. It take the same author to define the semantics for it to be a complete keyword.

Comment (gregsdennis) [reply jsonschema-core.md:70]: Okay.  Are you happy, then, to have `x-` keywords clarified as being "recognized" (using that word) in #1518?

File: jsonschema-core.md
Comment (jdesrosiers) [reply jsonschema-core.md:438]: I'm not convinced the rewording in #1518 really addresses this concern. Not enough anyway. I think that rewording here is probably a better approach. Something like,

> Implementations MUST refuse to evaluate schemas which contain keywords they do not know how to process.

I think that's clear, covers all the bases, and avoids confusion about whether `x-` counts as a recognized keyword. Implementations know how to process `x-` keywords, so it's clear this statement isn't referring to those keywords.

File: jsonschema-core.md
Comment (jdesrosiers) [reply jsonschema-core.md:70]: I think that would work, but I just posted in the other tread an alternative that I think would be better. It avoids the problem altogether.

File: jsonschema-core.md
Comment (gregsdennis) [reply jsonschema-core.md:438]: That also addresses implementations which choose to support custom keywords outside of vocabs.

Comment (gregsdennis) [reply jsonschema-core.md]: Updated with one tweak.

**Review by jdesrosiers — APPROVED** [review]
I still think that paragraph in the overview section should reference `x-` as an extension mechanism, but I'll let you decide if you want to make that change. Otherwise, this looks good to me.

**Comment by gregsdennis** [comment]
I'm going to leave it out of the extensions section.

> `x-foo` and `x-bar` have the same annotation collection and validation behavior, but they aren't the same keyword. They have different semantics given to them by the schema author, not by the spec.

You're not wrong about this.  However, those extra semantics only apply outside of the context of evaluating the schema, regardless of how or for what purpose the evaluation occurs.

The specification is saying that `x-` keywords are annotations only, which means the only semantics they carry or behavior they exhibit _during evaluation of the schema_ is collection of their values as annotations.  Any other semantics that a schema author or application infers _must_ be carried out outside of the context of evaluation.

File: jsonschema-core.md
Comment (Relequestual) [reply jsonschema-core.md:980]: I would be happier if there was an Issue which identifies the inconsistency which is part of the stable release milestone. I'm fine if #1510 resolves that Issue.
I'm just hesitent to approve a PR with a known contradiction **unless** there's a clear Issue to resolve it that likely won't be missed.

Comment (gregsdennis) [reply jsonschema-core.md:980]: There _is_ the discussion I linked.  It describes the problem in extreme detail, and I keep linking it in all the places that involves vocabs.

**Review by jviotti — APPROVED** [review]
This is my first time being involved in specs, so take my approval with a grain of salt, but at least all the proposed changes make perfect sense to me and it reads well. I think an implementer would definitely get it.

File: jsonschema-core.md
Comment (gregsdennis) [reply jsonschema-core.md:1002]: The discussion isn't going to get lost, and it doesn't need to be addressed before the spec releases, or even before #1510 merges.  It doesn't need to be tracked on a project board or be associated with a milestone.

Comment (Relequestual) [reply jsonschema-core.md:1002]: OK. I guess given #1510 is part of the milestone, I'm less concerned.

**Review by Relequestual — APPROVED** [review]

**Pull request merged**
